managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/PendingReadsLimiter.java

```diff
@@ -0,0 +1,126 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.bookkeeper.mledger.impl.cache;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.prometheus.client.Gauge;
+import lombok.AllArgsConstructor;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class PendingReadsLimiter {
+
+    private static final Gauge PULSAR_ML_PENDING_READS_BUFFER_SIZE = Gauge
+            .build()
+            .name("pulsar_ml_pending_reads_buffer_size")
+            .help("Estimated number of bytes for pending reads from storage")
+            .register();
+
+    private final long maxPendingReadsBufferSize;
+    private long remainingPendingRequestsBytes;
+
+    public PendingReadsLimiter(long maxPendingReadsBufferSize) {
+        this.maxPendingReadsBufferSize = maxPendingReadsBufferSize;
+        this.remainingPendingRequestsBytes = maxPendingReadsBufferSize;
+    }
+
+    @VisibleForTesting
+    public synchronized long getRemainingPendingRequestsBytes() {
+        return remainingPendingRequestsBytes;
+    }
+
+    @AllArgsConstructor
+    @ToString
+    static class Handle {
+        final long acquiredPermits;
+        final boolean success;
+        final int trials;
+
+        final long creationTime;
+    }
+
+    private static final Handle DISABLED = new Handle(0, true, 0, -1);
+
+    Handle acquire(long permits, Handle current) {
+        if (maxPendingReadsBufferSize <= 0) {
+            // feature is disabled
+            return DISABLED;
+        }
+        synchronized (this) {
```

Comment:
This could create additional contention in the read path. I don't see an immediate solution. I am only mentioning in case someone else has concerns. I assume this is why `maxMessagePublishBufferSizeInMB` is divided up to create limits for each thread.